src/fromager/models.py

```diff
@@ -0,0 +1,244 @@
+import logging
+import os
+import pathlib
+import string
+import typing
+
+import yaml
+from packaging.utils import NormalizedName, canonicalize_name
+from packaging.version import Version as BaseVersion
+from pydantic import (
+    BaseModel,
+    BeforeValidator,
+    ConfigDict,
+    Field,
+    GetCoreSchemaHandler,
+    HttpUrl,
+    ValidationInfo,
+    field_validator,
+)
+from pydantic_core import CoreSchema, core_schema
+
+logger = logging.getLogger(__name__)
+
+
+# build directory
+def _before_builddirectory(p: str) -> pathlib.Path:
+    result = pathlib.Path(p)
+    if os.pardir in str(result) or result.is_absolute():
+        raise ValueError("{result!r} is not a relative path")
+    return result
+
+
+BuildDirectory = typing.Annotated[
+    pathlib.Path,
+    BeforeValidator(_before_builddirectory),
+]
+
+
+# version
+class Version(BaseVersion):
+    """Pydantic-aware package version"""
+
+    @classmethod
+    def validate(cls, v: typing.Any, info: core_schema.ValidationInfo) -> BaseVersion:
+        if isinstance(v, BaseVersion):
+            return v
+        return BaseVersion(v)
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, source_type: typing.Any, handler: GetCoreSchemaHandler
+    ) -> CoreSchema:
+        return core_schema.with_info_plain_validator_function(
+            cls.validate,
+            serialization=core_schema.plain_serializer_function_ser_schema(
+                str, when_used="json"
+            ),
+        )
+
+
+# Changelog
+Changelog = dict[Version, list[str]]
+
+# environment variables map
+EnvVars = dict[str, str | int | float | bool]
+
+# Package validates and transforms name to canonicalized name
+Package = typing.Annotated[
+    NormalizedName,
+    BeforeValidator(lambda pkg: canonicalize_name(pkg, validate=True)),
+]
+
+# URL with templating
+UrlTemplate = typing.NewType("UrlTemplate", HttpUrl)
+
+# build variant
+Variant = typing.NewType("Variant", str)
+
+# common settings
+MODEL_CONFIG = ConfigDict(extra="forbid", use_attribute_docstrings=True)
+
+
+class ResolverDist(BaseModel):
+    """Packages resolver dist"""
+
+    model_config = MODEL_CONFIG
+
+    sdist_server_url: HttpUrl | None = None
+    """Source distribution download server"""
+
+    include_sdists: bool | None = None
+    """Include source distribution?"""
+
+    include_wheels: bool | None = None
+    """Include wheels?"""
+
+
+class DownloadSource(BaseModel):
+    """Package download source
+
+    Download package sources from an alternative source, e.g. GitHub release.
+    """
+
+    model_config = MODEL_CONFIG
+
+    url: UrlTemplate | None = None
+    """Source download url (string template)"""
+
+    destination_filename: str | None = None
```

Comment:
We want to make sure this is a valid filename, and has no path components. Maybe we can add that in a separate PR, once we have the basics in place.

Reply:
I'm adding a simple validator that checks for `os.pathsep not in value`